### PR TITLE
Add Perceptual Hash computation to imagehasher

### PR DESCRIPTION
**Pull request by mizaki**
This returns the same hex result (from my limit testing) as https://github.com/JohannesBuchner/imagehash/blob/master/imagehash/__init__.py#L260 (phash) but without using scipy or numpy.

### CODE REVIEW

File: comictaggerlib/imagehasher.py [hunk 1]
```diff
@@ -71,118 +72,100 @@ def set_bit(x: int, idx_val: tuple[int, int]) -> int:
 
         return result
 
-    def average_hash2(self) -> None:
```
Comment (lordwelch) [inline comictaggerlib/imagehasher.py]: Please return the average_hash2 function it should not be removed in this PR.
Was any of this code taken from another project? If so it needs to be stated here.

File: comictaggerlib/imagehasher.py [hunk 2]
```diff
@@ -71,118 +72,100 @@ def set_bit(x: int, idx_val: tuple[int, int]) -> int:
...
+
+        pixels = convert_image_to_ndarray(image)
+        dct = generate_dct2(generate_dct2(pixels, axis=0), axis=1)
+        dctlowfreq = [row[:8] for row in dct[:8]]
+        med = median([item for sublist in dctlowfreq for item in sublist])
+        # Convert to a bit string
+        diff = "".join(str(int(item > med)) for row in dctlowfreq for item in row)
+        # Convert to hex
+        width = int(math.ceil(len(diff) / 4))
+        result = "{:0>{width}x}".format(int(diff, 2), width=width)
```
Comment (lordwelch) [inline comictaggerlib/imagehasher.py]: A string should not be returned this should be an int.
```python
diff = [int(item > med) for row in dctlowfreq for item in row]
result = int(diff, 2)
```

File: comictaggerlib/imagehasher.py [hunk 3]
```diff
@@ -71,118 +72,100 @@ def set_bit(x: int, idx_val: tuple[int, int]) -> int:
...
+        elif isinstance(h1, str) and isinstance(h2, str):
             # convert hex strings to ints
             n1 = int(h1, 16)
             n2 = int(h2, 16)
+        else:
+            # Mixed hashes or some other problem so return a high number. Should return None instead?
+            return 999
 
         # xor the two numbers
-        n = n1 ^ n2
```
Comment (lordwelch) [inline comictaggerlib/imagehasher.py]: This typing was correct before please revert.

File: comictaggerlib/issueidentifier.py [hunk 4]
```diff
@@ -53,7 +53,7 @@ class SearchKeys(TypedDict):
 class Score(TypedDict):
     score: NotRequired[int]
     url: str
-    hash: int
```
Comment (lordwelch) [inline comictaggerlib/issueidentifier.py]: This should always be an int

File: comictaggerlib/issueidentifier.py [hunk 5]
```diff
@@ -132,11 +132,9 @@ def set_hasher_algorithm(self, algo: int) -> None:
     def set_output_function(self, func: Callable[[str], None]) -> None:
         self.output_function = func
 
-    def calculate_hash(self, image_data: bytes) -> int:
-        if self.image_hasher == 3:
-            return -1  # ImageHasher(data=image_data).dct_average_hash()
+    def calculate_hash(self, image_data: bytes) -> int | str:
         if self.image_hasher == 2:
-            return -1  # ImageHasher(data=image_data).average_hash2()
```
Comment (lordwelch) [inline comictaggerlib/issueidentifier.py]: Please revert and use `self.image_hasher == 3` for the phash function

File: comictaggerlib/issueidentifier.py [hunk 6]
```diff
@@ -280,7 +278,7 @@ def get_issue_cover_match_score(
         issue_id: str,
         primary_img_url: str,
         alt_urls: list[str],
-        local_cover_hash_list: list[int],
```
Comment (lordwelch) [inline comictaggerlib/issueidentifier.py]: This should always be an int

File: comictaggerlib/resulttypes.py [hunk 7]
```diff
@@ -10,7 +10,7 @@ class IssueResult(TypedDict):
     distance: int
     issue_number: str
     cv_issue_count: int | None
-    url_image_hash: int
```
Comment (lordwelch) [inline comictaggerlib/resulttypes.py]: This should always be an int

**Review by lordwelch — CHANGES_REQUESTED** [review]

File: comictaggerlib/imagehasher.py (hunk 1 above)
Comment (mizaki) [reply comictaggerlib/imagehasher.py]: It was made with the help of ChatGPT. I did a search before that and didn't find anything that I could make fit.

File: comictaggerlib/imagehasher.py (hunk 2 above)
Comment (mizaki) [reply comictaggerlib/imagehasher.py]: The `diff` change created an array which wouldn't work with `int`. I left `diff` as is to create a bit string and int'ed that.

**Review by lordwelch — APPROVED** [review]

**Pull request merged**